crates/worker/src/executor/parameter_server.rs

```diff
@@ -336,3 +303,150 @@ impl JobExecutor for ParameterServerExecutor {
         Ok(ParameterServerExecution { task_tracker })
     }
 }
+
+/// Applies a binary operation to corresponding tensors from two safetensor files.
+///
+/// This function memory-maps two input safetensor files, iterates through the tensors
+/// of the first file, finds the corresponding tensor by name in the second file,
+/// and applies the provided operation `op` to the pair. The resulting tensors
+/// are saved into a `temp_path` and combined into a single result file. Only
+/// two tensors will be held in memory at the same time.
+///
+/// # Arguments
+/// * `file_a_path` - Path to the first safetensor file.
+/// * `file_b_path` - Path to the second safetensor file.
+/// * `output_path` - Path where the resulting safetensor file will be saved.
+/// * `temp_path` - Path to a temporary directory where the intermediate safetensor file will be saved.
+/// * `device` - The Candle device to perform computations on (e.g., `Device::Cpu`).
+/// * `op` - A closure that takes two tensors and returns a new tensor.
+///
+/// # Returns
+/// A `Result` indicating success or a `TensorOpError` on failure.
+///
+/// # Note
+/// Tensors present in the second file but not the first are ignored. If a tensor
+/// from the first file is not found in the second, it is skipped with a warning.
+/// The `temp_path` will be created and delted by the function. Make sure it doesn't
+/// point to an existing directory that contains important data.
+/// Also make sure that the tensors in `op` are in the same order as they are passed to the function.
+async fn apply_tensor_op<F>(
+    file_a_path: &Path,
+    file_b_path: &Path,
+    output_path: &Path,
+    temp_path: &Path,
+    device: &Device,
+    op: F,
+) -> Result<(), TensorOpError>
+where
+    F: Fn(&Tensor, &Tensor) -> Result<Tensor, candle_core::Error>,
+{
+    // 1. Open both safetensor files in a memory-mapped way.
+    // SAFETY: The MmapedSafetensors::new function is unsafe because it assumes
+    // the underlying file will not be modified while the memory map is active.
+    let tensors_a = unsafe { candle_core::safetensors::MmapedSafetensors::new(file_a_path)? };
+    let tensors_b = unsafe { candle_core::safetensors::MmapedSafetensors::new(file_b_path)? };
+    fs::create_dir_all(temp_path).await?;
+
+    let mut result_tensors = Vec::new();
+
+    // 2. Iterate through each tensor in the first file.
+    for (name, tensor_view) in tensors_a.tensors() {
+        // Try to load the corresponding tensor from the second file.
+        match tensors_b.load(&name, device) {
+            Ok(tensor_b) => {
+                // If found, load the tensor from the first file.
+                let tensor_a = tensor_view.load(device)?;
+
+                // 3. Apply the provided computation function and serialize the result to disk.
+                let result_tensor = op(&tensor_a, &tensor_b)?;
```

Comment:
🤔 I wonder whether we may need/want to move this or the entire `apply_tensor_op`  to separate _blocking thread_ (or tokio yield_now()) to not starve  workers.